### ISSUE TITLE
Fuse RL loop into main execution

### ISSUE THREAD

**Issue by OliBus801** [OliBus801/TARL-simulator#1]
# Fuse RL loop into main execution (`main.py`)

## Summary

Unify the current split between `main.py` (classical algorithms like Dijkstra) and `train.py` (DirectionMPNN + PPO RL) into a single entry point. The goal is one `main.py` that can **select the learning algorithm** and **optionally train** via CLI flags (or config), avoiding duplicated loops and inconsistent code paths.

## Motivation

* Two separate loops make running experiments clunky and error‑prone.
* Duplicated setup (env/sim init, logging, seeding, device, I/O).
* Harder to compare methods and share infrastructure (checkpoints, metrics).

## Proposal

Create a unified runner that encapsulates:

* **Algorithm selection:** `--algo {dijkstra,mpnn,mpnn+ppo}`
* **Mode selection:** `--mode {eval,train}` (e.g., `train` only valid for RL)
* **Common args:** scenario, seeds, device, steps, timelimit, output paths
* **Training args (when applicable):** lr, batch size, epochs, PPO clip, gamma, GAE λ, rollout horizon, eval freq, early stop, checkpoints

### Sketch CLI

```bash
python main.py \
  --algo mpnn+ppo \
  --mode train \
  --scenario data/siouxfalls \
  --seed 42 --device cuda \
  --epochs 50 --rollout-steps 4096 \
  --ppo-clip 0.2 --gamma 0.99 --gae-lambda 0.95 \
  --chkpt-dir runs/exp1 --eval-every 5 \
  --log-wandb false
```

### Design

* Introduce `Runner` (or `Experiment`) with a uniform interface:

  * `setup()` → build `TransportationSimulator`, agents, `DirectionMPNN`, PPO, optimizers
  * `train_loop()` → RL loop only
  * `eval_loop()` → shared evaluation for all algos
* Move loop logic from `train.py` into reusable functions/classes under `rl/` (e.g., `ppo_trainer.py`).
* Keep agents pluggable (`mpnn_agent.py`, `transformer_agent.py`, classical/Dijkstra agent).
* Centralize metrics (road optimality, agent flow, V/C, etc.) and logging.
* Optional: support `--config path.yaml` to load defaults, with CLI overrides.

## Tasks

* [ ] Add `Runner` (unified setup, train, eval)
* [ ] Refactor PPO training from `train.py` → `rl/ppo_trainer.py`
* [ ] Refactor classical pathfinding callsite into `algorithms/dijkstra_runner.py`
* [ ] Implement CLI in `main.py` (argparse / tyro), with validation
* [ ] Centralize seeding, device placement, and reproducibility
* [ ] Centralize logging (stdout + file), metrics dump, and checkpoints
* [ ] Add YAML config loader (optional) with CLI override
* [ ] Update docs/README with new usage examples
* [ ] Add smoke tests:

  * [ ] `--algo dijkstra --mode eval` runs 1 short episode
  * [ ] `--algo mpnn+ppo --mode train` runs 1 short epoch and saves a checkpoint
  * [ ] `--algo mpnn --mode eval` runs inference with pretrained weights
* [ ] Deprecate `train.py` (keep stub that points to `main.py` for one release)

## Acceptance criteria (DoD)

* Single `main.py` can run:

  * Dijkstra evaluation
  * MPNN evaluation (no training)
  * MPNN+PPO training + evaluation
* Checkpoints saved & loadable; evaluation can consume them.
* Metrics and artifacts written to a user‑specified `--output-dir`.
* README updated; CI smoke tests pass on CPU.

## Notes / Integration points

* Reuse `TransportationSimulator`, `DirectionMPNN`, `FeatureHelpers`, and agent classes.
* Ensure metric computation is centralized (no duplicate implementations).
* Keep interfaces stable so adding future algos (e.g., Transformer agent) is trivial.

---
